nexus/src/app/sagas/region_snapshot_replacement_start.rs

```diff
@@ -0,0 +1,1128 @@
+            },
+            RegionAllocationParameters::FromRaw {
+                block_size: alloc_region_params.block_size,
+                blocks_per_extent: alloc_region_params.blocks_per_extent,
+                extent_count: alloc_region_params.extent_count,
+            },
+            &params.allocation_strategy,
+            alloc_region_params.current_allocated_regions.len() + 1,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(datasets_and_regions)
+}
+
+async fn rsrss_alloc_new_region_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let log = osagactx.log();
+
+    let alloc_region_params =
+        sagactx.lookup::<AllocRegionParams>("alloc_region_params")?;
+
+    let maybe_dataset_and_region = find_only_new_region(
+        log,
+        alloc_region_params.current_allocated_regions,
+        sagactx.lookup::<Vec<(db::model::Dataset, db::model::Region)>>(
+            "new_datasets_and_regions",
+        )?,
+    );
+
+    // It should be guaranteed that if rsrss_alloc_new_region succeeded then it
+    // would have bumped the region redundancy, so we should see something here.
+    // Guard against the case anyway.
+    if let Some(dataset_and_region) = maybe_dataset_and_region {
+        let (_, region) = dataset_and_region;
+        osagactx
+            .datastore()
+            .regions_hard_delete(log, vec![region.id()])
+            .await?;
+    }
+
+    Ok(())
+}
+
+async fn rsrss_find_new_region(
+    sagactx: NexusActionContext,
+) -> Result<(db::model::Dataset, db::model::Region), ActionError> {
+    let osagactx = sagactx.user_data();
+    let log = osagactx.log();
+
+    let alloc_region_params =
+        sagactx.lookup::<AllocRegionParams>("alloc_region_params")?;
+
+    let maybe_dataset_and_region = find_only_new_region(
+        log,
+        alloc_region_params.current_allocated_regions,
+        sagactx.lookup::<Vec<(db::model::Dataset, db::model::Region)>>(
+            "new_datasets_and_regions",
+        )?,
+    );
+
+    let Some(dataset_and_region) = maybe_dataset_and_region else {
+        return Err(ActionError::action_failed(Error::internal_error(
+            &format!(
+                "expected dataset and region, saw {:?}!",
+                maybe_dataset_and_region,
+            ),
+        )));
+    };
+
+    Ok(dataset_and_region)
+}
+
+async fn rsrss_new_region_ensure(
+    sagactx: NexusActionContext,
+) -> Result<
+    (nexus_db_model::Dataset, crucible_agent_client::types::Region),
+    ActionError,
+> {
+    let params = sagactx.saga_params::<Params>()?;
+    let osagactx = sagactx.user_data();
+    let log = osagactx.log();
+
+    // With a list of datasets and regions to ensure, other sagas need to have a
+    // separate no-op forward step for the undo action to ensure that the undo
+    // step occurs in the case that the ensure partially fails. Here this not
```

Comment:
Nit: final sentence:   "Here this is not required,.."

Reply:
nice, 73c95c6c7